rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,517 @@
+|-------------------|--------|------------------------------------------------------------|
+| `SUCCESS`         | `0x00` | No errors has been occured                                 |
+| `UNKNOWN_COMMAND` | `0x01` | The SDK does not support operation requested by the server |
+
+
+### `GET_TELEMETRY`
+
+The request must not have payload.
+
+The response might have payload if the SDK has metrics to report.
+
+The SDK must reset local counters after successful scheduling response for the
+network delivery.
+
+The format of the response follows the [Text-based exposition format of the
+Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format),
+and includes metrics and histograms collected by the SDK since the previous
+report.
+
+#### Metrics
+
+    sdk_kv_r_utimedout{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"} 1 1695747260
+
+Where
+
+* `sdk_kv_r_utimedout` is a name of the metric, one of the:
+
+  | Metric | Description |
+  | ---- | ---- |
+  | `sdk_{service}_r_utimedout` | Unambiguous timeout |
+  | `sdk_{service}_r_atimedout` | Ambiguous timeout |
+  | `sdk_{service}_r_canceled` | Canceled |
+  | `sdk_{service}_r_total` | Total number of operations |
+
+* `{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)", id="66388CF5BFCF7522/18CC8791579B567C", node="node1", bucket="travel-sample"}` is a list of the tags.
+
+  | Label | Description |
+  | ---- | ---- |
+  | `agent` | agent string of the SDK, as it is used everywhere (HELLO message, logs etc.) |
+  | `id` | SDK identifier string as specified by [SDK-RFC #35](0035-rto.md#client--connection-ids) in section "Client / Connection IDs" |
+  | `node` | The hostname of the node as seen in configuration, and used by the SDK to establish connection. |
+  | `alt_node` | If alternative address is used, this field should be present |
+  | `bucket` | Name of the bucket (if present) |
+
+* `1` is a metric value. Each metric value increment represents a single network operation (e.g. `GET_ALL_REPLICAS` should be counted as `1 + number_of_replicas`).
+* `1695747260` is a metric timestamp, that represents time in milliseconds from the Epoch (1970-01-01 00:00:00 UTC) to the moment of the generation this response.
+
+#### Histograms
+
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.001",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 24054
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.010",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 33444
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.100",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 100392
+    sdk_kv_retrieval_duration_seconds_bucket{le="0.5",agent="sdk/2.4.5.0",id="66388C/18CC8",bucket="travel-sample",node="node1"} 129389
```

Comment:
The 'le' bins need to be consistent across SDKs, so that they are aggregated together.
These examples show a way in which they could end up inconsistent if not properly considered: Some of the bins are given with 3 decimal places, including trailing zeros, and yet some only give 1 decimal place, without any trailing zeros. The SDKs will need to consistently either using a fixed number of decimal places, or have no trailing zeros. @avsej Please make it clear in the RFC whether the bins will use trailing zeros with fixed number of decimal places, or no trailing zeros.

Reply:
Thanks, I have updated the spec.